tutorials/best-practices/readme.md

```diff
@@ -0,0 +1,119 @@
+
+## Model training
+In some cases, you will define large models or need to use large pre-trained models. We recommend that you write the model defiition inside the `train` function. The reasoning similar to the one we just mentioned in the dataset section above. For example, when building deep learning models we recommend that you write the instructions to download images in the training fuction. Doing otherwise means that you will have to endure longer waiting time as the images are uploaded. Writing the download instructions in the train function downloads the images on the Layer infra and they are ready to use immediately. Here is a code snippet showing how you might download some images and extract them on Layer infra. 
+```python
+@pip_requirements(packages=["wget"])
+@fabric("f-gpu-small")
+@model(name="model_name")
+def train():
+    from tensorflow.keras.preprocessing.image import ImageDataGenerator
+    import tensorflow as tf
+    from tensorflow import keras
+    from tensorflow.keras import Sequential
+    from tensorflow.keras.layers import Dense,Conv2D,MaxPooling2D,Flatten,Dropout
+    from tensorflow.keras.preprocessing.image import ImageDataGenerator
+    from tensorflow.keras.callbacks import EarlyStopping
+    import os
+    import matplotlib.pyplot as plt 
+    from PIL import Image
+    import numpy as np
+    import pandas as pd
+    import tarfile
+    import wget
+    wget.download(url)
+    food_tar = tarfile.open('data.tar.gz')
+    food_tar.extractall('.') 
+    food_tar.close()
+    
+    base_dir = '...'
+    class_names = os.listdir(base_dir)
+    train_datagen = ImageDataGenerator(...)
+    validation_gen = ImageDataGenerator(rescale=1./255,validation_split=0.2)
+    image_size = (200, 200)
+    training_set = train_datagen.flow_from_directory(...)
+    validation_set = validation_gen.flow_from_directory(...)
+    model =......
+    model.compile()
+    epochs=10
+    history = model.fit(..)
+    metrics_df = pd.DataFrame(history.history)
+    layer.log({"Metrics":metrics_df})
+    loss, accuracy = model.evaluate(..)
+    layer.log({"Accuracy on test dataset":accuracy})
+    metrics_df[["loss","val_loss"]].plot()
+    layer.log({"Loss plot":plt.gcf()})
+    metrics_df[["categorical_accuracy","val_categorical_accuracy"]].plot()
+    layer.log({"Accuracy plot":plt.gcf()})
+    return model
+  layer.run([train])
+
+```
+## Declare dependencies 
+It is good practice to declare dependencies when building entites that depend on other Layer entities. This enables Layer to optimize your pipeline. You can declare dependencies for models and datasets. 
+
+
+```python
+from layer import Dataset, Model
+
+#MODEL DECORATOR WITH DEPENDENCIES
+@model("clustering_model",dependencies=[Dataset("product_ids_and_vectors")])
+
+#DATASET DECORATOR WITH DEPENDENCIES
+@dataset("final_product_clusters", dependencies=[Model("clustering_model"), Dataset("product_ids_and_vectors")])
+
+
```

Comment:
Let's add a section for Pip requirements. People should pass the exact version not to have any errors. Their local version of sklearn may not match what we have on fabrics. We can put a link to here:
https://docs.app.layer.ai/docs/reference/fabrics#preinstalled-libraries